cmd/snap/cmd_run.go

```diff
@@ -227,6 +230,145 @@ func isReexeced() bool {
 	return strings.HasPrefix(exe, dirs.SnapMountDir)
 }
 
+func migrateXauthority(info *snap.Info) (string, error) {
+	u, err := userCurrent()
+	if err != nil {
+		return "", fmt.Errorf(i18n.G("cannot get the current user: %s"), err)
+	}
+
+	// If our target directory (XDG_RUNTIME_DIR) doesn't exist we
+	// don't attempt to create it.
+	baseTargetDir := filepath.Join(dirs.XdgRuntimeDirBase, u.Uid)
+	if !osutil.FileExists(baseTargetDir) {
+		return "", nil
+	}
+
+	xauthPath := osGetenv("XAUTHORITY")
+	if len(xauthPath) == 0 || !osutil.FileExists(xauthPath) {
+		// Nothing to do for us. Most likely running outside of any
+		// graphical X11 session.
+		return "", nil
+	}
+
+	fin, err := os.Open(xauthPath)
+	if err != nil {
+		return "", err
+	}
+	defer fin.Close()
+
+	// Abs() also calls Clean(); see https://golang.org/pkg/path/filepath/#Abs
+	xauthPathAbs, err := filepath.Abs(fin.Name())
+	if err != nil {
+		return "", nil
+	}
+
+	// Remove all symlinks from path
+	xauthPathCan, err := filepath.EvalSymlinks(xauthPathAbs)
+	if err != nil {
+		return "", nil
+	}
+
+	// Verify XAUTHORITY matches the cleaned path
```

Comment:
I had a bit of trouble understanding what is going on here, i.e. why this is done etc. A small comment would be helpful I think. Maybe something like: `Ensure the XAUTHORITY env is not abused by checking that it point to exactly the file we just opened (no symlinks, no funny "../.." etc)`. Or a better command, maybe @jdstrand has some input.